vortex-ipc/src/iterator.rs

```diff
@@ -0,0 +1,183 @@
+        match reader.read_message()? {
+            Some(msg) => match msg {
+                DecoderMessage::DType(dtype) => Ok(SyncIPCReader { reader, ctx, dtype }),
+                msg => {
+                    vortex_bail!("Expected DType message, got {:?}", msg);
+                }
+            },
+            None => vortex_bail!("Expected DType message, got EOF"),
+        }
+    }
+}
+
+impl<R: Read> ArrayIterator for SyncIPCReader<R> {
+    fn dtype(&self) -> &DType {
+        &self.dtype
+    }
+}
+
+impl<R: Read> Iterator for SyncIPCReader<R> {
+    type Item = VortexResult<ArrayData>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.reader.read_message().transpose()? {
+            Ok(msg) => match msg {
+                DecoderMessage::Array(array_parts) => Some(
+                    array_parts
+                        .into_array_data(self.ctx.clone(), self.dtype.clone())
+                        .and_then(|array| {
+                            if array.dtype() != self.dtype() {
+                                Err(vortex_err!(
+                                    "Array data type mismatch: expected {:?}, got {:?}",
+                                    self.dtype(),
+                                    array.dtype()
+                                ))
+                            } else {
+                                Ok(array)
+                            }
+                        }),
+                ),
+                msg => Some(Err(vortex_err!("Expected Array message, got {:?}", msg))),
+            },
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+/// A trait for converting an [`ArrayIterator`] into an IPC stream.
+pub trait ArrayIteratorIPC {
+    fn into_ipc(self) -> ArrayIteratorIPCBytes
+    where
+        Self: Sized;
+
+    fn write_ipc<W: Write>(self, write: W) -> VortexResult<W>
+    where
+        Self: Sized;
+}
+
+impl<I: ArrayIterator + 'static> ArrayIteratorIPC for I {
+    fn into_ipc(self) -> ArrayIteratorIPCBytes
+    where
+        Self: Sized,
+    {
+        let mut encoder = MessageEncoder::new(ALIGNMENT);
+        let buffers = encoder.encode(EncoderMessage::DType(self.dtype()));
+        ArrayIteratorIPCBytes {
+            inner: Box::new(self),
+            encoder,
+            buffers,
+        }
+    }
+
+    fn write_ipc<W: Write>(self, mut write: W) -> VortexResult<W>
+    where
+        Self: Sized,
+    {
+        let mut stream = self.into_ipc();
+        for buffer in &mut stream {
+            write.write_all(buffer?.as_slice())?;
+        }
+        Ok(write)
+    }
+}
+
+pub struct ArrayIteratorIPCBytes {
+    inner: Box<dyn ArrayIterator + 'static>,
+    encoder: MessageEncoder,
+    buffers: Vec<Buffer>,
+}
+
+impl ArrayIteratorIPCBytes {
+    /// Collects the IPC bytes into a single `Vec<u8>` buffer.
+    pub fn collect_to_buffer(self) -> VortexResult<Vec<u8>> {
+        let mut buffer = vec![];
```

Comment:
this is why this PR is failing miri. You can't write this data into an unaligned Vec<u8> and then attempt to use it to back a PrimitiveArray